### PR TITLE
feat: show derived configuration

### PR DESCRIPTION
**Pull request by mcknasty**
# **[Pull Request 517](https://github.com/bcoe/c8/pull/517)**

## Commit Message

commit b33e8e7773ea79897c8b77f92894a7cb3fc5aaf7
Author: Trevor D McKeown <tmckeown@gmail.com>
Date:   Sun Feb 18 09:17:58 2024 -0500

    Print Derived Configuration Report (#517)

    feat: print derived config variables
    feat: print derived config variables in json
    test: 16 new unit tests to support features
    test: 1 skipped unit test for discovered bug in yargs with reports param

## **Following the Contributions Recommendations [here.](https://github.com/bcoe/c8/blob/main/CONTRIBUTING.md)**

1. [x] Make sure you have installed the latest version of Node.js
2. [x] Fork this project on GitHub and clone your fork locally
3. [x] Create local branches to work within. These should also be created directly from the main branch.  Local Fork [here.](https://github.com/mcknasty/c8)
4. [x] Make your changes.
5. [x] Run tests to make sure all is okay (everything should pass except the snapshot).
   1. A [complete log](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-initial-test-txt) of initial test results.
   2. As instructed, ignore snapshot failures. [1 pending](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-initial-test-txt-L172)
   3. [111 passing](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-initial-test-txt-L171) in [54 seconds](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-initial-test-txt-L171)
6. [x] Now update the snapshot.
   1. A [complete log](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-snapshot-test-txt) of snapshot test results.
   2. [111 passing](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-snapshot-test-txt-L175) in [54 seconds](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-snapshot-test-txt-L175)
7. [x] If all is passing, commit your changes. The log of the commit can be found [here.](https://github.com/bcoe/c8/pull/517/commits/32dd6afec6a19eb36c71a3189991775792207057)
8. [x] As a best practice, once you have committed your changes, it is a good idea to use git rebase (not git merge) to synchronize your work with the main repository.
9. [x] Run tests again to make sure all is okay.
    1. A [complete log](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-final-test-txt) of the final test results.
    2. [111 passing](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-final-test-txt-L171) in [55 seconds](https://gist.github.com/mcknasty/465e70521191d4ab96906d70fccf0b0d#file-final-test-txt-L171)
10. [x] Push
11. [x] Open the pull request, see details in the template.
12. [ ] Make any necessary changes after review.

### CODE REVIEW

**Comment by mcknasty** [comment]
Saw a few nit picks.  I might be able to refactor this patch to use fewer files, by naming the snapshots in the test cases.  WIP.

**Comment by mcknasty** [comment]
Ready to go!

File: test/print-config-helpers.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
+const { spawnSync } = require('child_process')
```
Comment (mcknasty) [inline test/print-config-helpers.js:1]: Change file name to helpers.js.

File: c8-ascii-art.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/* ________/\\\\\\\\\_        _____/\\\\\\\\\____         */
+/*  _____/\\\////////__        ___/\\\///////\\\__        */
+/*   ___/\\\/___________        __\/\\\_____\/\\\__       */
+/*    __/\\\_____________        __\///\\\\\\\\\/___      */
```
Comment (bcoe) [inline c8-ascii-art.txt:4]: I'd be tempted to just make this a template string rather than bothering to add a file.

File: lib/parse-args.js [hunk 3]
```diff
@@ -158,8 +158,17 @@ function buildYargs (withCommands = false) {
       describe: 'supplying --merge-async will merge all v8 coverage reports asynchronously and incrementally. ' +
         'This is to avoid OOM issues with Node.js runtime.'
     })
+    .options('print-config', {
+      default: false,
+      type: 'boolean',
+      describe: 'Print the derived configuration between command line parameters and loaded configuration file'
+    })
+    .options('print-config-format', {
+      default: 'text',
```
Comment (bcoe) [inline lib/parse-args.js:167]: You can use choices for this, to enforce `json` or `text`:

https://github.com/yargs/yargs/blob/main/docs/api.md#choiceskey-choices

**Review by bcoe — COMMENTED** [review]
Looks like a useful addition to functionality to me, I would change the PR title to:

`feat: show derived configuration`

**Comment by mcknasty** [comment]
Thanks @bcoe.  I will make the changes above and have a new commit in the next week.

**Comment by mcknasty** [comment]
@bcoe, Please review again.

File: lib/print-config.js [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    const cleanArgs = cleanUpArgumentObject(args)
+
+    if (args.printConfigFormat === 'text') {
+      printConfigText(cleanArgs, cmdExecuted)
+    } else if (checkArgs.printConfigFormat === 'json') {
+      const jsonYargs = JSON.stringify(cleanArgs, 2)
+      console.log(jsonYargs)
+    }
+
+    // DO NOT REMOVE!  This is intentional.
```
Comment (bcoe) [inline lib/print-config.js:42]: Could you expand on the comment to explain why this `process.exit()` is necessary?

Why is this `process.exit()` necessary?

File: test/test-helpers.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    ? [c8Path, args]
+    : [c8Path, ...args]
+
+  return runSpawn(nodePath, argsv, options)
+}
+
+/**
+ * Function: runSpawn
+ *
+ * @param {String} cmd: A string representing the prompt command
```
Comment (bcoe) [inline test/test-helpers.js:38]: This seems like quite a bit of test helper logic being introduced for the size of the feature.

Could we get away with simply using a regex assertion and testing against some of the expected output?

File: lib/print-config.js (hunk 4 above)
Comment (mcknasty) [reply lib/print-config.js:42]: The reason for this `process.exit()` is that it stops c8 from producing a coverage report after showing the configuration report.  Additionally, this logic allows JSON to be also printed, and possibly consumed by another process.

@bcoe can you anticipate any use cases where a configuration report and coverage report are necessary?

**Comment by bcoe** [comment]
@mcknasty it's mostly my fault for not being more engaged on this repo, but I think I need to declare bankruptcy on some of the inflight PRs.

If you're still feeling passionate about some of these changes (some were pretty cool IMO, like printing derived configuration), mind reopening on at a time, with a tracking issue that describes the functionality you'd like to see.

Thank you for the work you have done on this repo 👏

**Comment by mcknasty** [comment]
Thanks for your comments on these.  Some of the PRs did need to get closed.  Unfortunately, this one is a prerequisite for #518 at the moment.

**Pull request closed without merging**